src/commands.jl

```diff
@@ -0,0 +1,183 @@
+pauses at a breakpoint, the reference to the breakpoint is returned.
+"""
+function finish_and_return!(@nospecialize(recurse), frame::Frame, istoplevel::Bool=false)
+    pc = finish!(recurse, frame, istoplevel)
+    isa(pc, BreakpointRef) && return pc
+    return get_return(frame)
+end
+finish_and_return!(frame::Frame, istoplevel::Bool=false) = finish_and_return!(finish_and_return!, frame, istoplevel)
+
+"""
+    bpref = dummy_breakpoint(recurse, frame::Frame)
+
+Return a fake breakpoint. This can be useful as the `recurse` argument to `evaluate_call!`
+(or any of the higher-order commands) to ensure that you return immediately after stepping
+into a call.
+"""
+dummy_breakpoint(@nospecialize(recurse), frame::Frame) = BreakpointRef(frame.framecode, 0)
+
+"""
+    ret = finish_stack!(recurse, frame, istoplevel=false)
+    ret = finish_stack!(frame, istoplevel=false)
+
+Unwind the callees of `frame`, finishing each before returning to the caller.
+`frame` itself is also finished
+If execution hits a breakpoint, `ret` will be a reference to the breakpoint.
+"""
+function finish_stack!(@nospecialize(recurse), frame::Frame, istoplevel::Bool=false)
+    frame0 = frame
+    frame = leaf(frame)
+    while true
+        ret = finish_and_return!(recurse, frame, istoplevel)
+        isa(ret, BreakpointRef) && return ret
+        frame === frame0 && return ret
+        recycle(frame)
+        frame = caller(frame)
+        frame === nothing && return ret
+        pc = frame.pc
+        if isassign(frame, pc)
+            lhs = getlhs(pc)
+            do_assignment!(frame, lhs, ret)
+        end
+        pc += 1
+        frame.pc = pc
+        shouldbreak(frame) && return BreakpointRef(frame.framecode, pc)
+    end
+    return frame
+end
+finish_stack!(frame::Frame, istoplevel::Bool=false) = finish_stack!(finish_and_return!, frame, istoplevel)
+
+"""
+    next_until!(predicate, recurse, frame, istoplevel=false)
+    next_until!(predicate, frame, istoplevel=false)
+
+Execute the current statement. Then step through statements of `frame` until the next
+statement satifies `predicate(stmt)`.
+"""
+function next_until!(@nospecialize(predicate), @nospecialize(recurse), frame::Frame, istoplevel::Bool=false)
+    pc = step_expr!(recurse, frame, istoplevel)
+    while pc !== nothing && !isa(pc, BreakpointRef)
+        if predicate(pc_expr(frame, pc)) || shouldbreak(frame, pc)
+            return pc
+        end
+        pc = step_expr!(recurse, frame, istoplevel)
+    end
+    return pc
+end
+next_until!(predicate, frame::Frame, istoplevel::Bool=false) =
+    next_until!(predicate, finish_and_return!, frame, istoplevel)
+
+next_call!(@nospecialize(recurse), frame::Frame, istoplevel::Bool=false) =
+    next_until!(is_call_or_return, recurse, frame, istoplevel)
+next_call!(frame::Frame, istoplevel::Bool=false) = next_call!(finish_and_return!, frame, istoplevel)
+
+function maybe_next_call!(@nospecialize(recurse), frame::Frame, istoplevel::Bool=false)
```

Comment:
```suggestion
"""
    maybe_next_call!(predicate, frame, istoplevel=false)

Return the current statement of `frame` if it is a `:return` or `:call` expression. Otherwise, step through the statements of `frame` until the next `:return` or `:call` expression.
"""
function maybe_next_call!(@nospecialize(recurse), frame::Frame, istoplevel::Bool=false)
```